File: 数据结构与算法之美/MergeSort.java
package 数据结构与算法之美;

/**
 * @author wang xiao
 * @date 2022/6/23
 */
public class MergeSort {

    public static void main(String[] args) {
        int [] a = new int[]{3,4,0,1,5,6,7,8};
        MergeSort mergeSort = new MergeSort();
        mergeSort.mergeSort(a,0,a.length);

        for (int i : a) {
            System.out.println(i);
        }
    }

    public void mergeSort (int[] arr,int low,int high){
        while (low <high){
            int mid = (low+high)/2;
            mergeSort(arr,low,mid);
            mergeSort(arr,mid+1,high);
            merge(arr,low,mid,high);
        }
    }

    public void merge(int [] arr,int left,int mid, int right){
        int [] temp = new int[arr.length];
        int i = left;
        int j = mid+1;
        int index = left;

        while (i<= mid && j<= right){
            if (arr[i] <=  arr[j]){
                temp[index++] = arr[i++];
            }else {
                temp[index++] = arr[j++];
            }
        }

        while (i <= mid){
            temp[index++]= arr[i++];
        }
        while (j<right){
            temp[index++] = arr[j++];
        }
        for (int c=left;c<=right;c++){
            arr[c] = temp[c];
        }
    }
}

File: Go/datastruct/merge_sort.go
package datastruct

func MergeSort(arr []int, left, right int) {

	if left < right {
		mid := (left + right) / 2
		MergeSort(arr, left, mid)
		MergeSort(arr, mid+1, right)
		merge(arr, left, mid, right)
	}
}
func merge(arr []int, left, mid, rigtht int) {
	temp := make([]int, rigtht-left+1)

	i, j, index := left, mid+1, 0

	for i < mid && j < rigtht {
		if arr[i] < arr[j] {
			temp[index] = arr[i]
			i++
		} else {
			temp[index] = arr[j]
			j++
		}
		index++
	}

	for i < mid {
		temp[index] = arr[i]
		index++
		i++
	}

	for j < rigtht {
		temp[index] = arr[j]
		index++
		j++
	}

	for i := left; i <= rigtht; i++ {
		arr[i] = temp[i]
	}

}
